[ENT-628] [ENT-629] [ENT-630] [ENT-631] [ENT-634] Display Program Landing Page Details

UmanShahzad opened the pull request:
**Description:** This PR adds the remaining program details on the program enrollment landing page that we need. This includes:

- "What is `PLATFORM_NAME`?" on the sidebar, with a configurable `PLATFORM_DESCRIPTION`.
- Program type description on the sidebar, with the program type logo.
- Configurable `ENTERPRISE_TAGLINE` on the page header for our landing pages plus our data sharing consent page.
- A "What you'll learn" section, that is expandable for when there are multiple expected learning outcomes.
- A "Real Career Impact" section, that is expandable for when there are multiple corporate endorsements.
- Courses on the course list in the program summary box have pacing information below them.
- Courses on the course list in the program summary box can be clicked to bring up our course modal (the same one from the course enrollment landing page).
- Price data, as well as a discount provider, is written in the "Price" row of the program summary box.

**JIRA:** [ENT-628](https://openedx.atlassian.net/browse/ENT-628), [ENT-629](https://openedx.atlassian.net/browse/ENT-629), [ENT-630](https://openedx.atlassian.net/browse/ENT-630), [ENT-631](https://openedx.atlassian.net/browse/ENT-631), & [ENT-634](https://openedx.atlassian.net/browse/ENT-634).

**Merge deadline:** End of sprint.

**Testing instructions:**

### Setup

1. Make authoring organizations on course-discovery.
1. Make multiple corporate endorsements on course-discovery.
1. Make courses (audit and verified track) and course runs, with filled out information for most fields, and use any authoring organization if applicable. Make them appear on course-discovery.
1. Make a program type -- one of `Professional Certificate`, `MicroMasters Certificate`, or `XSeries Certificate`.
1. Make a program containing anything you created above, if applicable (and note the UUID). Make it one-click purchasable, and add expected learning outcomes.
1. Make a program API config (actually, it's just a "program config", but we aren't changing the name yet on the model) at LMS path `/admin/programs/programsapiconfig/`. Make sure it's enabled.
1. Make SKU data associated with the course(s) you made above on ecommerce.
1. Make an enterprise customer. Note the UUID.

### Testing Program Details

1. Login to the LMS as any user who is not enrolled in the courses & program you created.
1. Navigate to `/enterprise/{enterprise_uuid}/program/{program_uuid}/enroll/`. Replace `{enterprise_uuid}` and `{program_uuid}` with the enterprise's UUID and the program's UUID that you noted above, respectively.
1. Make sure you see the following information (depending on your configuration):

#### Tagline

![selection_005](https://user-images.githubusercontent.com/10018065/30351666-f202bff8-9835-11e7-9fbb-72c78b57c7d5.png)

#### Sidebar

![selection_007](https://user-images.githubusercontent.com/10018065/30351668-f63defca-9835-11e7-93fa-4460456c364c.png)

#### What you'll learn (unexpanded)

![selection_008](https://user-images.githubusercontent.com/10018065/30351678-0545e86a-9836-11e7-830e-8af5e4773b29.png)

#### What you'll learn (expanded)

![selection_009](https://user-images.githubusercontent.com/10018065/30351684-0ac59bf0-9836-11e7-9519-998e38f536eb.png)

#### Real Career Impact (unexpanded)

![selection_010](https://user-images.githubusercontent.com/10018065/30351692-13c98bb2-9836-11e7-9026-a070f00d6ff3.png)

#### Real Career Impact (expanded)

![selection_011](https://user-images.githubusercontent.com/10018065/30351700-1a7f5144-9836-11e7-9a9e-b04089546b81.png)

#### Program Summary Box (course pacing, clickable course titles, and discount provider)

![selection_012](https://user-images.githubusercontent.com/10018065/30351703-29641b2c-9836-11e7-91d6-58f05de8dcb8.png)

#### Course Modal (initial open, course 1)

![selection_013](https://user-images.githubusercontent.com/10018065/30351710-334d6152-9836-11e7-907e-5f77b1bae097.png)

#### Course Modal (scrolled down completely, course 1)

![selection_014](https://user-images.githubusercontent.com/10018065/30351717-3c044752-9836-11e7-8c9d-156705d4852a.png)

#### Course Modal (initial open, any other course, i.e. course 2)

![selection_015](https://user-images.githubusercontent.com/10018065/30351729-46516b0e-9836-11e7-8271-b86c02f7e511.png)

**Merge checklist:**

- [x] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [x] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=186"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

UmanShahzad commented on enterprise/views.py:
```diff
@@ -1009,32 +1059,41 @@ def get_program_details(self, request, program_uuid):
         if program_details is None:
             raise Http404
 
+        program_type_slug = program_details['type'].lower().replace(' ', '-')
```
Just realized I should be using `django.utils.text.slugify` here. Will do it on the next push after initial code review.

UmanShahzad commented on tests/test_enterprise/views/test_program_enrollment_view.py:
```diff
@@ -840,3 +973,21 @@ def test_post_program_enrollment_view_redirect_to_basket(
+    @mock.patch('enterprise.views.CourseCatalogApiServiceClient')
+    def test_extend_course_no_course_details_returned_from_discovery(self, course_catalog_api_client_mock):
+        """
+        We raise a 404 when there are Discovery API returns no course details in ``extend_course``.
```
Note to self: fix this bad copy pasta doc string.

UmanShahzad commented:
@douglashall @haikuginger @brittneyexline Note -- although I've done my best to separate the commits out corresponding to their tickets, I found it easier for myself to read through it one go. Of course, use the style best suited for you :)

UmanShahzad commented:
Ugh, so sorry. I lost the code that adds CSS for the new sidebar changes (very stupid reason)... -_- It wasn't much luckily, but I'll have to redo it again and deploy it to the sandbox.

haikuginger commented on enterprise/constants.py:
```diff
@@ -28,3 +28,24 @@
 # Course mode sorting based on slug
 COURSE_MODE_SORT_ORDER = ['verified', 'professional', 'no-id-professional', 'audit', 'honor']
+
+PROGRAM_TYPE_DESCRIPTION = {
```
Should these details be configurable somehow? For example, to be able to add blurbs for other program types.

UmanShahzad replied:
They were originally supposed to be through a new `description` field on discovery, but that got scratched. CC @douglashall for more details.

UmanShahzad commented:
@haikuginger Thank you for your review! I'll try it myself with a screen reader, and if I notice anything we can easily add / fix, I'll do that. But yeah, @cptvitamin is probably the best at this.

cptvitamin commented:
@UmanShahzad do you have a sandbox?  I'd be happy to review.

UmanShahzad commented:
@cptvitamin Yep :) Sent on chat.

brittneyexline commented on enterprise/views.py:
```diff
@@ -988,19 +980,77 @@ class ProgramEnrollmentView(NonAtomicView):
+            'course_full_description': clean_html_for_template_rendering(course_run_details['full_description'] or ''),
+            'expected_learning_items': course_details.get('expected_learning_items', []),
+            'staff': course_run_details.get('staff', []),
+            'course_effort': ungettext_min_max(
```
should we put this effort calculation in its own function for readability? i see we're doing the same thing above.

UmanShahzad replied:
There's already the `ungettext_min_max` abstraction I put on top of it. We could do another but any ideas where it should live? Maybe a new Mixin class for these views?

UmanShahzad commented:
@brittneyexline Yep they are now inconsistent side panels, but all I did from my side was copy the mock. So @chris-mike might have more comments there.

I agree about the sidebar panel, I didn't get a chance to see a logo yet on the sidebar because of the upload issues and my own local devstack issues. I'll add some margin/padding there.

Yeah the corporate endorsement logos are pretty large, but all I've done is limit their height to 100px or so, which is I believe the same as the side panel logo. I could control width but yeah, that'd distort the image and potentially look really bad. This comes down more to the customers who send the logos in the first place; they should be sending proportions that'll make their logos look nice on the page.

UmanShahzad commented:
The latest version of this is now on the sandbox.

UmanShahzad commented:
@brittneyexline Spacing between the logo has been added and deployed to the sandbox.

UmanShahzad commented:
Accessibility review has been addressed. Thanks a bunch @cptvitamin !!

UmanShahzad commented on enterprise/static/enterprise/enterprise_program_enrollment_page.css:
```diff
@@ -1,8 +1,13 @@
+        display: block;
+    }
+}
+
```
p.s. this isn't actually a part of the accessibility review, sorry..

UmanShahzad commented:
@haikuginger Just to note, some more accessibility stuff came up which I'm in the middle of addressing at the moment. Mobile responsiveness (a few nits) also needs to be covered, but that'll be right after accessibility.

UmanShahzad commented on tox.ini:
```diff
@@ -62,21 +62,12 @@ deps =
-    pylint --py3k tests
-    pylint --py3k test_utils
+    pylint -j 0 enterprise consent integrated_channels tests test_utils
+    pylint -j 0 --py3k enterprise consent integrated_channels tests test_utils
```
This runs much faster on my machine now.

UmanShahzad commented:
Rebased from `759774ecef8913a737181c9d11bacf5df8f390b3` on `master`.

UmanShahzad commented:
The latest version of this is now on the sandbox (minus mobile responsiveness changes -- that's next).

chris-mike commented:
We cautioned of overly large corporate endorsement logos in the acceptance criteria. If you could set logo height to 30 px, that would be about right.

I'm not concerned with inconsistent sidebars at this point. Thanks!

UmanShahzad commented:
@chris-mike That's absolutely fine, yep. Will do that.

UmanShahzad commented on enterprise/templates/enterprise/enterprise_program_enrollment_page.html:
```diff
@@ -85,7 +85,7 @@ <h3 class="program-metadata-header-text">{{ expected_learning_items_text }}</h3>
                 </ul>
                 {% if expected_learning_items|length > expected_learning_items_show_count %}
-                  {% expand_button see_more_text '#what-youll-learn' %}
+                  {% expand_button see_more_text 'what-youll-learn' %}
```
This commit would be better labeled "Remove `#` from aria-control" :) sorry!

UmanShahzad commented:
Rebased on `master` from 30f27dca70bd6d94a7600353cb5edb447d2a5809.

UmanShahzad commented:
@chris-mike Ready for mobile responsiveness review!

UmanShahzad commented:
@brittneyexline @douglashall UI/UX review is done -- just need a final from you guys.

P.S. We shouldn't merge this just yet, of course, until #198 is merged and bumped on the edx-platform.

UmanShahzad commented:
I'll be merging this today, but still needs a changelog, etc..

UmanShahzad commented:
Rebased on `master` from 1eee20f523eafaceebb85e7761a69ff7c856a772 and added changelog/version bump.

I will be merging 5 commits because they correspond to separate tickets.
